pkg/asset/installconfig/ibmcloud/metadata.go

```diff
@@ -92,6 +101,50 @@ func (m *Metadata) SetCISInstanceCRN(crn string) {
 	m.cisInstanceCRN = crn
 }
 
+// DNSInstance returns a DNSInstance holding information about the DNS Services instance
+// managing the DNS zone for the base domain.
+func (m *Metadata) DNSInstance(ctx context.Context) (*DNSInstance, error) {
+	if m.dnsInstance != nil {
+		return m.dnsInstance, nil
+	}
+
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	// Prevent multiple attempts to retrieve (set) the dnsInstance if it hasn't been set (multiple threads reach mutex concurrently)
+	if m.dnsInstance == nil {
+		client, err := m.Client()
+		if err != nil {
+			return nil, err
+		}
+
+		zones, err := client.GetDNSZones(ctx, types.InternalPublishingStrategy)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, z := range zones {
+			if z.Name == m.BaseDomain {
+				if z.InstanceID == "" || z.InstanceCRN == "" {
+					return nil, fmt.Errorf("dnsInstance has unknown ID/CRN: %q - %q", z.InstanceID, z.InstanceCRN)
+				}
+				m.SetDNSInstance(z.InstanceID, z.InstanceCRN)
+				return m.dnsInstance, nil
+			}
+		}
+		return nil, fmt.Errorf("dnsInstance unknown due to DNS zone %q not found", m.BaseDomain)
+	}
+	return m.dnsInstance, nil
+}
+
+// SetDNSInstance sets the DNS Services instance with the instance ID and instance CRN
+func (m *Metadata) SetDNSInstance(instanceID string, instanceCRN string) {
+	m.dnsInstance = &DNSInstance{
+		ID:  instanceID,
+		CRN: instanceCRN,
+	}
+}
+
```

Comment:
is the purpose of the mutex in DNSInstance to guard against concurrent modification of m.dnsInstance?  if so, could this public function possibly negate that protection?

Reply:
That's a good point, although the public function allows access to test the `metadata` functions, perhaps I can work out another method for performing the testing of the functions in metadata without having to have public "set" functions.

Reply:
I removed the SetDNSInstance function for now, will handle necessary changes in upcoming unit test PR for metadata.